android/java/res/layout/fragment_fingerprint_unlock.xml

```diff
@@ -0,0 +1,78 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.core.widget.NestedScrollView
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:scrollbars="vertical"
+    android:fadeScrollbars="false"
+    android:background="@color/container_background"
+    android:fillViewport="true">
+
+    <RelativeLayout
```

Comment:
We should try not to use RelativeLayout. it's slower compared to other layouts. it wouldn't make a huge different for this layout but let's try to use different VIewGroup if possible.

Reply:
> We should try not to use RelativeLayout. it's slower compared to other layouts.

I tend to disagree with this statement; over the years I've heard it a lot, and most of the time the reason for such a claim is that relative layouts do two layout passes.

Doing two layout passes is now a negligible computation, so to me the decision of whether using a relative layout depends on several factors.

I agree that, for example, nested relative layouts should be avoided, otherwise the layout passes become exponential, and that may harm the performance.

There's a talk from Romain Guy (the original author of relative layouts) where they delve into a more detailed explanaition, including also some dos and donts: https://stackoverflow.com/a/17496262.

The takeaway from the talk is that for a complex hierarchy, doing an extra measure pass could potentially become costly, so using nest `RelativeLayouts`, produces an exponential measurement algorithm.

Let's apply this to `fragment_fingerprint_unlock` layout: the relative layout hierarchy is simple and totally flat, no extra view groups, just two layout passes and the refresh rate is at its maximum and perfectly stable.

Leveraging the `alignParentBottom` option, the two buttons in the lower part stick to the bottom, so the layout is also nice on foldables, tablets and big screens in general, here is a preview.

![Screenshot from 2024-06-24 17-10-13](https://github.com/brave/brave-core/assets/3308503/acbcaf78-ec94-42bf-b56d-5e0ce9b4c8b0)

![Screenshot from 2024-06-24 17-10-33](https://github.com/brave/brave-core/assets/3308503/55563833-a5ed-4f60-9bd2-cba424c9008f)

![Screenshot from 2024-06-24 17-11-02](https://github.com/brave/brave-core/assets/3308503/089dec8c-3d88-4180-ac4f-5aeb25d0167d)

To obtain the same result without using a relative layout I see two possible options:

- Linear layout using weights with a nested frame layout. So two layout passes again, and a less readable hierarchy. Not a real improvement.
- Constraint layout, this is probably the only contender that would make sense to consider, it would flatten the hierarchy and would let achieve a similar result. The biggest downside is: without a fully working IDE integration where we can use the Design view to implement the constraint layouts I usually prefer to introduce them only for complex hierarchies where the performance improvement is tangible. I found constraint layouts less intuitive, and unfortunately, a nightmare to maintain without Android Studio fully integrated.

If there are other options to suggest, I'm super happy to try.

Let me know what you think about my analysis.